e2e/alertmanager_test.go

```diff
@@ -20,188 +20,172 @@ import (
 				LocalObjectReference: corev1.LocalObjectReference{
 					Name: "my-secret-name",
 				},
 				Key: "my-secret-key",
 			},
 			ExternalURL: "https://alertmanager.mycompany.com/",
-		},
-		&corev1.Secret{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      "my-secret-name",
-				Namespace: tctx.pubNamespace,
-			},
-			Data: map[string][]byte{
-				"my-secret-key": []byte(alertmanagerConfig),
-			},
-		},
-		"my-secret-key",
-	)
-}
-
-func testAlertmanager(ctx context.Context, t *OperatorContext, spec *monitoringv1.ManagedAlertmanagerSpec, secret *corev1.Secret, key string) {
-	t.createOperatorConfigFrom(ctx, monitoringv1.OperatorConfig{
-		Collection: monitoringv1.CollectionSpec{
-			ExternalLabels: map[string]string{
-				"external_key": "external_val",
-			},
-			Filter: monitoringv1.ExportFilters{
-				MatchOneOf: []string{
-					"{job='foo'}",
-					"{__name__=~'up'}",
-				},
-			},
-			KubeletScraping: &monitoringv1.KubeletScraping{
-				Interval: "5s",
-			},
-		},
-		ManagedAlertmanager: spec,
-	})
-	t.Run("deployed", t.subtest(func(ctx context.Context, t *OperatorContext) {
-		t.Parallel()
-		testAlertmanagerDeployed(ctx, t, spec)
-	}))
-	t.Run("config set", t.subtest(func(ctx context.Context, t *OperatorContext) {
-		t.Parallel()
-		testAlertmanagerConfig(ctx, t, secret, key)
-	}))
-}
-
-func testAlertmanagerDeployed(ctx context.Context, t *OperatorContext, config *monitoringv1.ManagedAlertmanagerSpec) {
-	var err error
-	pollErr := wait.PollUntilContextTimeout(ctx, 3*time.Second, 2*time.Minute, true, func(ctx context.Context) (bool, error) {
-		var ss appsv1.StatefulSet
-		if err = t.Client().Get(ctx, client.ObjectKey{Namespace: t.namespace, Name: operator.NameAlertmanager}, &ss); err != nil {
-			if apierrors.IsNotFound(err) {
-				return false, nil
-			}
-			return false, fmt.Errorf("getting alertmanager StatefulSet failed: %w", err)
 		}
 
-		// Assert we have the expected annotations.
-		wantedAnnotations := map[string]string{
-			"components.gke.io/component-name":               "managed_prometheus",
-			"cluster-autoscaler.kubernetes.io/safe-to-evict": "true",
-		}
-		if diff := cmp.Diff(wantedAnnotations, ss.Spec.Template.Annotations); diff != "" {
-			return false, fmt.Errorf("unexpected annotations (-want, +got): %s", diff)
+		// Update OperatorConfig.
+		_, err = opClient.MonitoringV1().OperatorConfigs(operator.DefaultPublicNamespace).Update(ctx, config, metav1.UpdateOptions{})
+		if err != nil {
+			t.Fatalf("update operatorconfig: %s", err)
 		}
 
-		// If config spec is empty, no need to assert EXTRA_ARGS.
-		if config == nil {
-			return true, nil
-		}
-
-		const containerName = "alertmanager"
-		// TODO(pintohutch): clean-up wantArgs init logic.
-		var wantArgs []string
-		for _, c := range ss.Spec.Template.Spec.Containers {
-			if c.Name != containerName {
-				continue
+		err = wait.PollUntilContextCancel(ctx, pollDuration, false, func(ctx context.Context) (bool, error) {
```

Comment:
suggestion: consider a refactor to pull out the func that gets polled here. This gets to be a lot of nesting, especially once you get to the conditionals inside the `for` loop, inside this closure, inside another closure, inside the test helper function.

Reply:
The `wait.PollUntilContextCancel` function signature basically requires a closure, as the `wait.ConditionWithContextFunc` is a closure that only takes in a `context.Context`. Given we almost always want to fetch something from the `kubeClient`, the variable would need to be declared outside of the function call, leading to this pattern.

Given this function is provided by the k8s apimachinery library, I'm going to leave this pattern as-is. If we want to leverage a different utility in the future for our k8s tests, we can discuss then.